hasher-matcher-actioner/scripts/listener.py

```diff
@@ -44,24 +49,51 @@ def do_POST(self):
             "headers": str(self.headers),
             "body": str(post_data.decode("utf-8")),
         }
-        with Handler.threadLock:
-            Handler.post_counter += 1
+        with _Handler.count_lock:
+            _Handler.post_counter += 1
         self._set_response()
         self.wfile.write(json.dumps(payload).encode("utf-8"))
 
 
-def start_listening_web_server(
-    hostname: str = "localhost",
-    port: int = 8080,
-    handler=Handler,
-) -> ThreadingHTTPServer:
-    with Handler.threadLock:
-        Handler.post_counter = 0
-    web_server = ThreadingHTTPServer((hostname, port), handler)
-    server_thread = threading.Thread(target=web_server.serve_forever)
-    server_thread.daemon = True
-    server_thread.start()
-    return web_server
+class Listener:
+    def __init__(self) -> None:
+        self.web_server: t.Optional[ThreadingHTTPServer] = None
+
+    def start_listening(
+        self,
+        hostname: str = "localhost",
+        port: int = 8080,
+    ):
+        if self.web_server:
+            print(
+                "Listener already started\nTo reset counter call `stop_listening` first."
+            )
+
+        with _Handler.count_lock:
+            _Handler.post_counter = 0
+        self.web_server = ThreadingHTTPServer((hostname, port), _Handler)
+        server_thread = threading.Thread(target=self.web_server.serve_forever)
+        server_thread.daemon = True
+        server_thread.start()
+        self.server_url = f"http://{hostname}:{port}"
+        print(f"Listener server started {self.server_url}")
+
+    def stop_listening(self):
+        if self.web_server:
+            self.web_server.shutdown()
+            print("Listener server stopped")
+            self.web_server = None
+        else:
+            print("Listener server not found")
+
+    def get_post_request_count(self) -> int:
+        if not self.web_server:
+            print("Warning: Listener server not found")
+            return -1
+        # Could use a queue instead of a lock+count and use queue.qsize()
+        # to avoid blocking the handler when getting the count
```

Comment:
Probably overkill and the queue would grow very large by the time you are done. Also, what makes you think qsize will not take a lock? [1]

1: https://github.com/python/cpython/blob/main/Lib/queue.py#L92-L95

Reply:
yeah, this comment was from my memory of using queues awhile ago. 
The reality is not as simplistic as use queue to avoid locking but we could have a thread listening to via .get instead of manual sleep and check loop. I'll remove the comment for now to avoid confusion.